numpy/core/src/npymath/halffloat.c

```diff
@@ -416,21 +418,26 @@ npy_uint16 npy_doublebits_to_halfbits(npy_uint64 d)
             npy_set_floatstatus_underflow();
         }
 #endif
-        d_sig >>= (1009 - d_exp);
+        /*
+         * Unlike floats, doubles has enough room to shift left to align
+         * significand. This changes the constant compared to normal branch.
+         */
+        assert(d_exp - 998 >= 0);
+        d_sig <<= (d_exp - 998);
         /* Handle rounding by adding 1 to the bit beyond half precision */
 #if NPY_HALF_ROUND_TIES_TO_EVEN
         /*
          * If the last bit in the half significand is 0 (already even), and
          * the remaining bit pattern is 1000...0, then we do not add one
          * to the bit after the half significand.  In all other cases, we do.
          */
-        if ((d_sig&0x000007ffffffffffULL) != 0x0000020000000000ULL) {
-            d_sig += 0x0000020000000000ULL;
+        if ((d_sig&0x003fffffffffffffULL) != 0x0010000000000000ULL) {
```

Comment:
Might be more readable to express these as `(1 << n) - 1` and `(1 << n)` for some `n`. Fine to leave as it too.

Reply:
Yeah, should change it everywhere then though probably.

Reply:
Probably a good idea, these bit patterns are a error prone :/, so I may do that but not super soon maybe. For now I had already fixed it up and added a test for the one actual problem.

Reply:
spaces needed around `&`

Reply:
Without spaces is par for the course for this file - I don't think we need to block this PR on that.